### ISSUE TITLE
Bundle size and build pipeline issues

### ISSUE THREAD

**Issue by pilifs** [venasolutions/react-lazy-paginated-tree#22]
I'm going to use this issue as a way to track some ongoing stuff. I'll start making separate issues once these are done. See below for more details on each item.

- [x] https://github.com/venasolutions/react-lazy-paginated-tree/pull/24 webpack is not running in production mode
- [x] webpack-bundle-analyzer cannot parse index.js
- [x] https://github.com/venasolutions/react-lazy-paginated-tree/pull/30 Material UI is not a peer dependency
- [x] Theming decision / discussion (Material UI or framework agnostic)?
- [x] https://github.com/venasolutions/react-lazy-paginated-tree/pull/29 Move post CSS work from runtime to buildtime
- [x] https://github.com/venasolutions/react-lazy-paginated-tree/pull/30 Only load functionality we need from utility libraries
- [x] https://github.com/venasolutions/react-lazy-paginated-tree/pull/31 More aggressive bundle size limit
- [x] https://github.com/venasolutions/react-lazy-paginated-tree/pull/31 Add build instructions to README

***

### https://github.com/venasolutions/react-lazy-paginated-tree/pull/24 DONE: webpack is not running in production mode

Changing the mode from development to production immediately reduces the build output by 3.2mb (!). I'm going to make a PR now to fix this now with a command line argument.

We will probably want to split up the config more (dev / prod), especially if separate builds for different themes is the way you'd like to go (see below).

---

### DONE: webpack-bundle-analyzer cannot parse index.js

Working after upgrading to latest version of webpack-bundle-analyzer (3.03). Easiest way to analyze the bundle, from project root:

```
npm install -g webpack-bundle-analyzer
npm run build:profile
webpack-bundle-analyzer buildstats.json ./dist
```

---

### https://github.com/venasolutions/react-lazy-paginated-tree/pull/30 DONE: Material UI is not a peer dependency

This must be a peer dependency for it to work in Material UI apps because there are issues with loading multiple versions of Material UI on the same page. A simple fix for this is to have multiple package.json files, one at root for building and one inside src including the peer dependency, and a slightly more involved process to compile `src/index.js`. See this project for an example: https://github.com/dmtrKovalenko/material-ui-pickers

---

### DONE: Make a decision on theming (project can't build without Material UI)

See Bowei's post below. We're going to move all of the Material UI stuff into an optional theme that you'll be able to import from main entry point. The default will use a CSS file that you can import and override as necessary.

--

Right now this is essentially operating as a Material UI component. The tradeoff here is ease of setup / fully styled and working out of box vs. appealing to a broader development audience. The best of both worlds may be theming with Material UI as the first theme, which is going to take a bit of work to think through and setup.

I'm totally cool if you'd prefer to keep it as a Material UI component, but I think a decision needs to be made. :) This will affect the way the build and code is structured in a big way.

---

### https://github.com/venasolutions/react-lazy-paginated-tree/pull/29 DONE: Move CSS work to build time

We are bundling a bunch of stuff right now that we don't need to. `postcssJs` and `autoprefixer` seem to be the biggest offenders. We should be able to do all of this stuff with dev dependencies at build time rather than runtime.

---

### https://github.com/venasolutions/react-lazy-paginated-tree/pull/30 DONE: Only load functions we need from utility libraries

All of core-js is being imported into the project by `@babel/polyfill`, which is not usually included in bundles (and also, a different version of babel than our build).

---

### https://github.com/venasolutions/react-lazy-paginated-tree/pull/31 DONE: Set a realistic bundle size limit

The limit is set to 15kb gzipped & minified for now.

**Comment by pilifs** [comment]
webpack bundle analyzer works now after upgrading to 3.03. I added a profile command in my open PR. The easiest way to run is to install the latest `webpack-bundle-analyzer` globally, then from project root do:

```
npm run build:profile
webpack-bundle-analyzer buildstats.json ./dist
```

Here's the what it looks like:

![image](https://user-images.githubusercontent.com/3997407/48305570-20062d00-e4fb-11e8-85b5-0fca8b1af6d1.png)

caniuse is loaded by autoprefixer... let's start there :) I'm also not sure why react-dom is still around. It shouldn't be. I see it in peerDependencies.

**Comment by pilifs** [comment]
Here's the latest bundle analyzer output after #24 and #25

![image](https://user-images.githubusercontent.com/3997407/48307051-babe3600-e512-11e8-806e-f92ecb8a4763.png)

Total size is now 342kb uncompressed. We can probably do way better than 100kb :)

**Comment by boweihan** [comment]
Per the theme discussion, it seems like there are a few items that need to be addressed in order to break the material-ui dependency, allow for easier overrides and reduce the bundle size:

1 - make material-ui a peer dependency - this will make it easier to override components and avoid version conflicts (lerna/rollup to manage the build)
2 - switch all material-ui styling to use withStyles() which is much smaller than the currently imported autoprefixer
3 - (long term) make the default theme use pure css. All current themes use material-ui.

**Comment by pilifs** [comment]
@don-p 

#1 depends on the library. In Material UI, multiple versions in the same bundle can cause a conflict with their class name generator. See https://material-ui.com/getting-started/faq/#why-aren--39-t-my-components-rendering-correctly-in-production-builds- . There are some workarounds on that page but you shouldn't have to deal with them.

#2 I agree with you. We'll do that soon.

**Comment by pilifs** [comment]
https://github.com/venasolutions/react-lazy-paginated-tree/pull/30 reduces the bundle size to 51kb uncompressed / 13kb gzipped.

Latest webpack bundle analyzer:

![image](https://user-images.githubusercontent.com/3997407/48326629-54b3da80-e608-11e8-86f0-1c21d7ed9967.png)

There's potentially more to be done (using async / await functions alone takes up ~15kb) but I'm satisfied. It's a pretty significant improvement over [where we started](https://github.com/venasolutions/react-lazy-paginated-tree/blob/407e5afca5fa45f1f27da0fece28334039c933e9/dist/index.js) before this weekend. :) After we merge #30, I'll update the README with instructions and add a size limit test, then we can close this.

**Comment by pilifs** [comment]
All done :) I'm going to close this now. Because some of the changes to peer dependencies will break projects that relied on the previously included Material UI dep, I added the 0.8.0 label and made a corresponding branch for that upcoming release: https://github.com/venasolutions/react-lazy-paginated-tree/tree/0.8.0

@boweihan Let's chat about what else you'd like to include in 0.8
